File: crates/ferrogate-payments/src/intent_test.rs
// Token4AI Cloud Attribution
// Developed by the commercial cloud service company represented by https://token4ai.cloud.
// Author: jamesduan (X: https://x.com/JamesDuanL)
// Created: 2026-07-26
// description: Unit tests for the immutable x402 payment-intent contract
// (issue #351): the method + request-body binding that makes a GET and a POST
// of a different body to the same URL distinguishable, sealed construction,
// validation rejections, the deterministic intent hash, and the challenge
// binding check. Sibling test file per AGENTS.md (no inline `mod tests {}`).

use super::*;

use base64::Engine as _;

use crate::wire::{parse_payment_required, select_requirement, RequirementFilter};

const USDC_DEVNET: &str = "4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU";
const CAIP2_DEVNET: &str = "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1";
const MERCHANT: &str = "2wKupLR9q6wXYppw8Gr2NvWxKBUqm4PPJKkQfoxHDBg4";
const OTHER_MERCHANT: &str = "9WzDXwBbmkg8ZTbNMqUxvQRAyrZzDsGYdLVL9zYtAWWM";
const FEE_PAYER: &str = "EwWqGE4ZFKLofuestmU4LDdK7XM1N4ALgdZccwYugwGd";
const RESOURCE: &str = "https://pay.example.com/weather";

fn selected(atomic: u64, recipient: &str) -> SelectedPayment {
    let body = serde_json::json!({
        "x402Version": 2,
        "resource": {"url": RESOURCE, "mimeType": "application/json"},
        "accepts": [{
            "scheme": "exact",
            "network": CAIP2_DEVNET,
            "amount": atomic.to_string(),
            "asset": USDC_DEVNET,
            "payTo": recipient,
            "maxTimeoutSeconds": 120,
            "extra": {"feePayer": FEE_PAYER}
        }]
    });
    let header = base64::engine::general_purpose::STANDARD.encode(body.to_string());
    let required = parse_payment_required(&header).expect("fixture challenge parses");
    select_requirement(&required, &RequirementFilter::default()).expect("fixture requirement")
}

fn identity() -> PaymentIntentIdentity {
    PaymentIntentIdentity {
        tenant_id: "tenant-1".into(),
        project_id: Some("project-1".into()),
        workspace_id: Some("workspace-1".into()),
        key_id: Some("key-1".into()),
        run_id: Some("run-1".into()),
        worker_id: Some("worker-1".into()),
        request_id: "req-1".into(),
    }
}

fn intent(method: &str, body: &[u8]) -> PaymentIntent {
    PaymentIntent::from_selected(
        &selected(2_500, MERCHANT),
        method,
        RESOURCE,
        body,
        identity(),
    )
    .expect("fixture intent is valid")
}

/// The gap this whole type exists to close: without the method + body binding,
/// an authorized `GET` and a `POST` of a different body to the SAME url carry
/// identical payment terms and are indistinguishable to anything downstream.
#[test]
fn a_get_and_a_post_of_a_different_body_to_the_same_url_are_distinguishable() {
    let read = intent("GET", &[]);
    let write = intent("POST", br#"{"drain":"everything"}"#);

    // Same merchant terms, same URL, same challenge...
    assert_eq!(read.challenge_hash_hex(), write.challenge_hash_hex());
    assert_eq!(
        read.authorized_resource_url(),
        write.authorized_resource_url()
    );
    assert_eq!(read.atomic_amount(), write.atomic_amount());
    // ...and yet the two intents are not the same intent.
    assert_ne!(read.http_method(), write.http_method());
    assert_ne!(read.request_body_hash(), write.request_body_hash());
    assert_ne!(
        read.intent_hash_hex(),
        write.intent_hash_hex(),
        "the intent hash must separate two different requests paying one challenge"
    );
}

#[test]
fn the_same_request_always_produces_the_same_intent_hash() {
    let first = intent("POST", b"body");
    let second = intent("post", b"body");

    // The method is normalized, so casing cannot fork the audit key.
    assert_eq!(second.http_method(), "POST");
    assert_eq!(first.intent_hash_hex(), second.intent_hash_hex());
    assert_eq!(first.intent_hash_hex().len(), 64);
    assert!(first
        .intent_hash_hex()
        .bytes()
        .all(|byte| byte.is_ascii_hexdigit()));
}

#[test]
fn changing_any_bound_field_changes_the_intent_hash() {
    let base = intent("GET", &[]);
    let baseline = base.intent_hash_hex();

    let other_amount =
        PaymentIntent::from_selected(&selected(2_501, MERCHANT), "GET", RESOURCE, &[], identity())
            .expect("valid");
    let other_recipient = PaymentIntent::from_selected(
        &selected(2_500, OTHER_MERCHANT),
        "GET",
        RESOURCE,
        &[],
        identity(),
    )
    .expect("valid");
    let other_url = PaymentIntent::from_selected(
        &selected(2_500, MERCHANT),
        "GET",
        "https://pay.example.com/other",
        &[],
        identity(),
    )
    .expect("valid");
    let mut other_run = identity();
    other_run.run_id = Some("run-2".into());
    let other_identity =
        PaymentIntent::from_selected(&selected(2_500, MERCHANT), "GET", RESOURCE, &[], other_run)
            .expect("valid");

    for (what, other) in [
        ("amount", other_amount),
        ("recipient", other_recipient),
        ("resource url", other_url),
        ("run id", other_identity),
    ] {
        assert_ne!(
            baseline,
            other.intent_hash_hex(),
            "a different {what} must produce a different intent hash"
        );
    }
}

/// An absent optional id and an empty-string one must not hash alike; the
/// latter is rejected outright, which is the stronger guarantee.
#[test]
fn an_absent_optional_id_is_not_the_same_as_a_blank_one() {
    let mut absent = identity();
    absent.run_id = None;
    let without_run =
        PaymentIntent::from_selected(&selected(2_500, MERCHANT), "GET", RESOURCE, &[], absent)
            .expect("an absent run id is legitimate");
    assert!(without_run.identity().run_id.is_none());
    assert_ne!(
        without_run.intent_hash_hex(),
        intent("GET", &[]).intent_hash_hex()
    );

    let mut blank = identity();
    blank.run_id = Some("   ".into());
    let error =
        PaymentIntent::from_selected(&selected(2_500, MERCHANT), "GET", RESOURCE, &[], blank)
            .expect_err("a blank run id is a caller bug, not an absence");
    assert!(matches!(
        error,
        PaymentIntentError::InvalidIdentity {
            field: "run_id",
            ..
        }
    ));
}

#[test]
fn an_intent_binds_to_exactly_one_challenge() {
    let bound = intent("GET", &[]);
    assert_eq!(bound.binding_mismatch(&selected(2_500, MERCHANT)), None);
    // A challenge for a different amount or payee is a different payment, and
    // the mismatch names the field so the caller can emit a specific reason.
    assert_eq!(
        bound.binding_mismatch(&selected(9_999, MERCHANT)),
        Some("challenge_hash")
    );
    assert_eq!(
        bound.binding_mismatch(&selected(2_500, OTHER_MERCHANT)),
        Some("challenge_hash")
    );
}

/// The binding check must not depend solely on the challenge hash: a caller
/// that hands over a hash-matching but field-tampered pair still fails.
#[test]
fn a_tampered_payment_term_is_caught_even_when_the_challenge_hash_matches() {
    let bound = intent("GET", &[]);
    let mut tampered = selected(2_500, MERCHANT);
    tampered.atomic_amount = 2_500_000;
    assert_eq!(bound.binding_mismatch(&tampered), Some("atomic_amount"));

    let mut tampered = selected(2_500, MERCHANT);
    tampered.recipient = OTHER_MERCHANT.to_string();
    assert_eq!(bound.binding_mismatch(&tampered), Some("recipient"));

    let mut tampered = selected(2_500, MERCHANT);
    tampered.mint = OTHER_MERCHANT.to_string();
    assert_eq!(bound.binding_mismatch(&tampered), Some("mint"));
}

#[test]
fn a_bodyless_request_has_a_concrete_hash_not_an_absence() {
    assert_eq!(RequestBodyHash::empty(), RequestBodyHash::of(&[]));
    assert_ne!(RequestBodyHash::empty(), RequestBodyHash::of(b"{}"));
    assert_eq!(RequestBodyHash::empty().as_hex().len(), 64);
    // Round-trips through the wire form.
    let hash = RequestBodyHash::of(b"payload");
    assert_eq!(RequestBodyHash::from_hex(&hash.as_hex()), Ok(hash));
    assert!(RequestBodyHash::from_hex("deadbeef").is_err());
    assert!(RequestBodyHash::from_hex(&hash.as_hex().to_uppercase()).is_err());
}

#[test]
fn construction_rejects_every_field_a_decision_would_later_depend_on() {
    let base = || PaymentIntentDraft {
        x402_version: 2,
        scheme: "exact".into(),
        network_caip2: CAIP2_DEVNET.into(),
        mint: USDC_DEVNET.into(),
        atomic_amount: 2_500,
        recipient: MERCHANT.into(),
        authorized_resource_url: RESOURCE.into(),
        http_method: "GET".into(),
        request_body_hash: RequestBodyHash::empty(),
        challenge_hash_hex: selected(2_500, MERCHANT).challenge_hash_hex(),
        max_timeout_seconds: 120,
        identity: identity(),
    };
    assert!(PaymentIntent::new(base()).is_ok());

    let cases: Vec<(&str, PaymentIntentDraft)> = vec![
        (
            "version",
            PaymentIntentDraft {
                x402_version: 1,
                ..base()
            },
        ),
        (
            "scheme",
            PaymentIntentDraft {
                scheme: "upto".into(),
                ..base()
            },
        ),
        (
            "network",
            PaymentIntentDraft {
                network_caip2: "solana:whatever".into(),
                ..base()
            },
        ),
        // A token symbol is never a mint.
        (
            "mint",
            PaymentIntentDraft {
                mint: "USDC".into(),
                ..base()
            },
        ),
        (
            "recipient",
            PaymentIntentDraft {
                recipient: "not-base58!".into(),
                ..base()
            },
        ),
        (
            "zero amount",
            PaymentIntentDraft {
                atomic_amount: 0,
                ..base()
            },
        ),
        (
            "relative url",
            PaymentIntentDraft {
                authorized_resource_url: "/weather".into(),
                ..base()
            },
        ),
        (
            "non-http url",
            PaymentIntentDraft {
                authorized_resource_url: "file:///etc/passwd".into(),
                ..base()
            },
        ),
        (
            "blank method",
            PaymentIntentDraft {
                http_method: "  ".into(),
                ..base()
            },
        ),
        (
            "smuggled method",
            PaymentIntentDraft {
                http_method: "GET /other HTTP/1.1".into(),
                ..base()
            },
        ),
        (
            "challenge hash",
            PaymentIntentDraft {
                challenge_hash_hex: "abc".into(),
                ..base()
            },
        ),
        (
            "zero timeout",
            PaymentIntentDraft {
                max_timeout_seconds: 0,
                ..base()
            },
        ),
        (
            "oversize timeout",
            PaymentIntentDraft {
                max_timeout_seconds: 86_401,
                ..base()
            },
        ),
        (
            "blank tenant",
            PaymentIntentDraft {
                identity: PaymentIntentIdentity {
                    tenant_id: " ".into(),
                    ..identity()
                },
                ..base()
            },
        ),
        (
            "blank request id",
            PaymentIntentDraft {
                identity: PaymentIntentIdentity {
                    request_id: String::new(),
                    ..identity()
                },
                ..base()
            },
        ),
    ];
    for (what, draft) in cases {
        assert!(
            PaymentIntent::new(draft).is_err(),
            "an invalid {what} must not produce an intent"
        );
    }
}

/// Deserialization must go through the same validation as construction, or a
/// persisted-then-reloaded intent could be weaker than a freshly built one.
#[test]
fn deserialization_cannot_bypass_validation() {
    let original = intent("POST", b"payload");
    let encoded = serde_json::to_string(&original).expect("intent serializes");
    let decoded: PaymentIntent = serde_json::from_str(&encoded).expect("round-trips");
    assert_eq!(decoded, original);
    assert_eq!(decoded.intent_hash_hex(), original.intent_hash_hex());
    // The atomic amount survives serialization as an integer, losslessly.
    let value: serde_json::Value = serde_json::from_str(&encoded).expect("json");
    assert_eq!(value["atomic_amount"], serde_json::json!(2_500u64));

    let tampered = encoded.replace(r#""scheme":"exact""#, r#""scheme":"upto""#);
    assert_ne!(
        tampered, encoded,
        "the fixture must actually be tampered with"
    );
    serde_json::from_str::<PaymentIntent>(&tampered)
        .expect_err("an unsupported scheme must not deserialize into a valid intent");

    let tampered = encoded.replace(r#""atomic_amount":2500"#, r#""atomic_amount":0"#);
    assert_ne!(tampered, encoded);
    serde_json::from_str::<PaymentIntent>(&tampered)
        .expect_err("a zero amount must not deserialize into a valid intent");
}

File: packages/payments/test/intent.test.ts
import { describe, expect, test } from "vitest";

import {
  challengeHashHex,
  parsePaymentRequired,
  PaymentIntent,
  PaymentIntentError,
  type PaymentIntentIdentity,
  RequestBodyHash,
  type SelectedPayment,
  selectRequirement,
} from "../src/index.js";
import { CAIP2_DEVNET, FEE_PAYER, OTHER_MERCHANT, PAY_TO, RESOURCE, USDC_DEVNET, toHeader } from "./fixtures.js";

const enc = new TextEncoder();

function selected(atomic: number, recipient: string): SelectedPayment {
  return selectedBig(BigInt(atomic), recipient);
}

/**
 * The same builder over a `bigint`, so a `u64` amount past 2^53 can be
 * constructed at all: `selected(Number(...))` would have rounded the amount in
 * the TEST before the code under test ever saw it, which is exactly the vacuity
 * the outbound guard exists to prevent. The amount travels through the
 * `PAYMENT-REQUIRED` header as a DECIMAL STRING, so nothing here loses digits.
 */
function selectedBig(atomic: bigint, recipient: string): SelectedPayment {
  const doc = {
    x402Version: 2,
    resource: { url: RESOURCE, mimeType: "application/json" },
    accepts: [
      {
        scheme: "exact",
        network: CAIP2_DEVNET,
        amount: atomic.toString(),
        asset: USDC_DEVNET,
        payTo: recipient,
        maxTimeoutSeconds: 120,
        extra: { feePayer: FEE_PAYER },
      },
    ],
  };
  return selectRequirement(parsePaymentRequired(toHeader(doc)));
}

function identity(): PaymentIntentIdentity {
  return {
    tenantId: "tenant-1",
    projectId: "project-1",
    workspaceId: "workspace-1",
    keyId: "key-1",
    runId: "run-1",
    workerId: "worker-1",
    requestId: "req-1",
  };
}

function intent(method: string, body: Uint8Array): PaymentIntent {
  return PaymentIntent.fromSelected(selected(2500, PAY_TO), method, RESOURCE, body, identity());
}

/**
 * LANGUAGE LIMIT PIN — kept as a PORT-TODO on `atomic_amount` in `src/intent.ts`.
 *
 * `u64` money survives as `bigint` everywhere in the in-memory domain; only the
 * serde hop is number-domain, because `JSON.parse` has already flattened the
 * integer to a double before `fromWire` is handed the value. These assertions
 * hold the boundary exactly where the marker says it is: if someone "fixes" the
 * marker by turning the domain type into `number`, the money path silently
 * becomes float and this fails.
 */
describe("PORT-TODO PIN — the money domain is bigint, the serde hop is not", () => {
  test("every in-memory atomic amount is a bigint, never a number", () => {
    const sel = selected(2500, PAY_TO);
    expect(typeof sel.atomicAmount).toBe("bigint");
    const built = PaymentIntent.fromSelected(sel, "GET", RESOURCE, new Uint8Array(0), identity());
    expect(typeof built.atomicAmount()).toBe("bigint");
    expect(built.atomicAmount()).toBe(2500n);
    // …and it survives a wire round trip as a bigint, not as whatever
    // `JSON.parse` handed back.
    const round = PaymentIntent.fromWire(JSON.parse(JSON.stringify(built.toWire())));
    expect(typeof round.atomicAmount()).toBe("bigint");
    expect(round.atomicAmount()).toBe(2500n);
    expect(round.intentHashHex()).toBe(built.intentHashHex());
  });

  /**
   * OUTBOUND HALF — CLOSED. `toWire()` used to write `Number(#atomicAmount)`
   * and round in silence past 2^53. These fail if that bare cast comes back.
   *
   * The third assertion is why this is a correctness gate and not cosmetics:
   * `intentHashHex()` digests the EXACT bigint, so a rounded wire amount makes
   * a verifier that re-derives the hash from the wire form disagree with the
   * signer, and `equals()` (JSON of `toWire()`) reports two DIFFERENT amounts
   * as one intent.
   */
  test("toWire REFUSES an amount past 2^53 instead of rounding it", () => {
    const over = BigInt(Number.MAX_SAFE_INTEGER) + 1n;
    const built = PaymentIntent.fromSelected(
      selectedBig(over, PAY_TO),
      "GET",
      RESOURCE,
      new Uint8Array(0),
      identity(),
    );
    // The domain accepted it — a u64 amount is legal on chain…
    expect(built.atomicAmount()).toBe(over);
    // …the WIRE render is what refuses.
    try {
      built.toWire();
      throw new Error("expected throw");
    } catch (error) {
      expect(error).toBeInstanceOf(PaymentIntentError);
      expect((error as PaymentIntentError).kind).toBe("amount_unrepresentable");
      expect((error as PaymentIntentError).field).toBe("atomic_amount");
    }
    // The exact boundary itself must still render.
    const edge = PaymentIntent.fromSelected(
      selectedBig(BigInt(Number.MAX_SAFE_INTEGER), PAY_TO),
      "GET",
      RESOURCE,
      new Uint8Array(0),
      identity(),
    );
    expect(edge.toWire().atomic_amount).toBe(Number.MAX_SAFE_INTEGER);
  });

  test("the refusal is what stops a rounded wire amount from breaking the hash binding", () => {
    const a = BigInt(Number.MAX_SAFE_INTEGER) + 1n;
    const b = BigInt(Number.MAX_SAFE_INTEGER) + 2n;
    // Two DIFFERENT on-chain amounts collapse onto ONE double…
    expect(a).not.toBe(b);
    expect(Number(a)).toBe(Number(b));
    const mk = (amount: bigint) =>
      PaymentIntent.fromSelected(
        selectedBig(amount, PAY_TO),
        "GET",
        RESOURCE,
        new Uint8Array(0),
        identity(),
      );
    // …but the intent hash, digested from the exact bigint, does NOT.
    expect(mk(a).intentHashHex()).not.toBe(mk(b).intentHashHex());
    // So emitting the collapsed number would sign one amount and publish
    // another. Both renders refuse.
    for (const amount of [a, b]) {
      expect(() => mk(amount).toWire()).toThrow(/amount_unrepresentable|losing precision/);
    }
    // …and `equals()`, which compares JSON of `toWire()`, cannot report them
    // as the same intent, because it cannot produce that JSON at all.
    expect(() => mk(a).equals(mk(b))).toThrow();
  });

  test("the number-domain hop is exactly the serde field, and it is exact below 2^53", () => {
    const wire = PaymentIntent.fromSelected(
      selected(2500, PAY_TO),
      "GET",
      RESOURCE,
      new Uint8Array(0),
      identity(),
    ).toWire();
    expect(typeof wire.atomic_amount).toBe("number");
    expect(wire.atomic_amount).toBe(2500);
    expect(Number.isSafeInteger(wire.atomic_amount)).toBe(true);
  });
});

describe("method + body binding (#351)", () => {
  test("a GET and a POST of a different body to the same URL are distinguishable", () => {
    const read = intent("GET", new Uint8Array(0));
    const write = intent("POST", enc.encode('{"drain":"everything"}'));
    expect(read.challengeHashHex()).toBe(write.challengeHashHex());
    expect(read.authorizedResourceUrl()).toBe(write.authorizedResourceUrl());
    expect(read.atomicAmount()).toBe(write.atomicAmount());
    expect(read.httpMethod()).not.toBe(write.httpMethod());
    expect(read.requestBodyHash().equals(write.requestBodyHash())).toBe(false);
    expect(read.intentHashHex()).not.toBe(write.intentHashHex());
  });

  test("the same request always produces the same intent hash (method normalized)", () => {
    const first = intent("POST", enc.encode("body"));
    const second = intent("post", enc.encode("body"));
    expect(second.httpMethod()).toBe("POST");
    expect(first.intentHashHex()).toBe(second.intentHashHex());
    expect(first.intentHashHex().length).toBe(64);
    expect(/^[0-9a-f]{64}$/.test(first.intentHashHex())).toBe(true);
  });

  test("changing any bound field changes the intent hash", () => {
    const baseline = intent("GET", new Uint8Array(0)).intentHashHex();
    const otherAmount = PaymentIntent.fromSelected(
      selected(2501, PAY_TO),
      "GET",
      RESOURCE,
      new Uint8Array(0),
      identity(),
    );
    const otherRecipient = PaymentIntent.fromSelected(
      selected(2500, OTHER_MERCHANT),
      "GET",
      RESOURCE,
      new Uint8Array(0),
      identity(),
    );
    const otherUrl = PaymentIntent.fromSelected(
      selected(2500, PAY_TO),
      "GET",
      "https://pay.example.com/other",
      new Uint8Array(0),
      identity(),
    );
    const otherRun = PaymentIntent.fromSelected(selected(2500, PAY_TO), "GET", RESOURCE, new Uint8Array(0), {
      ...identity(),
      runId: "run-2",
    });
    for (const other of [otherAmount, otherRecipient, otherUrl, otherRun]) {
      expect(baseline).not.toBe(other.intentHashHex());
    }
  });

  test("an absent optional id differs from a blank one (which is rejected)", () => {
    const withoutRun = PaymentIntent.fromSelected(
      selected(2500, PAY_TO),
      "GET",
      RESOURCE,
      new Uint8Array(0),
      { ...identity(), runId: null },
    );
    expect(withoutRun.identity().runId).toBeNull();
    expect(withoutRun.intentHashHex()).not.toBe(intent("GET", new Uint8Array(0)).intentHashHex());

    let err: unknown;
    try {
      PaymentIntent.fromSelected(selected(2500, PAY_TO), "GET", RESOURCE, new Uint8Array(0), {
        ...identity(),
        runId: "   ",
      });
    } catch (e) {
      err = e;
    }
    expect(err).toBeInstanceOf(PaymentIntentError);
    expect((err as PaymentIntentError).kind).toBe("invalid_identity");
    expect((err as PaymentIntentError).field).toBe("run_id");
  });
});

describe("binding check", () => {
  test("an intent binds to exactly one challenge", () => {
    const bound = intent("GET", new Uint8Array(0));
    expect(bound.bindingMismatch(selected(2500, PAY_TO))).toBeNull();
    expect(bound.bindingMismatch(selected(9999, PAY_TO))).toBe("challenge_hash");
    expect(bound.bindingMismatch(selected(2500, OTHER_MERCHANT))).toBe("challenge_hash");
  });

  test("a tampered payment term is caught even when the challenge hash matches", () => {
    const bound = intent("GET", new Uint8Array(0));
    const tamperedAmount = { ...selected(2500, PAY_TO), atomicAmount: 2500000n };
    expect(bound.bindingMismatch(tamperedAmount)).toBe("atomic_amount");
    const tamperedRecipient = { ...selected(2500, PAY_TO), recipient: OTHER_MERCHANT };
    expect(bound.bindingMismatch(tamperedRecipient)).toBe("recipient");
    const tamperedMint = { ...selected(2500, PAY_TO), mint: OTHER_MERCHANT };
    expect(bound.bindingMismatch(tamperedMint)).toBe("mint");
  });
});

describe("RequestBodyHash", () => {
  test("a bodyless request has a concrete hash, not an absence", () => {
    expect(RequestBodyHash.empty().equals(RequestBodyHash.of(new Uint8Array(0)))).toBe(true);
    expect(RequestBodyHash.empty().equals(RequestBodyHash.of(enc.encode("{}")))).toBe(false);
    expect(RequestBodyHash.empty().asHex().length).toBe(64);
    const hash = RequestBodyHash.of(enc.encode("payload"));
    expect(RequestBodyHash.fromHex(hash.asHex()).equals(hash)).toBe(true);
    expect(() => RequestBodyHash.fromHex("deadbeef")).toThrow();
    expect(() => RequestBodyHash.fromHex(hash.asHex().toUpperCase())).toThrow();
  });
});

describe("construction and validation", () => {
  const baseDraft = () => ({
    x402Version: 2,
    scheme: "exact",
    networkCaip2: CAIP2_DEVNET,
    mint: USDC_DEVNET,
    atomicAmount: 2500n,
    recipient: PAY_TO,
    authorizedResourceUrl: RESOURCE,
    httpMethod: "GET",
    requestBodyHash: RequestBodyHash.empty(),
    challengeHashHex: challengeHashHex(selected(2500, PAY_TO)),
    maxTimeoutSeconds: 120,
    identity: identity(),
  });

  test("the base draft is valid", () => {
    expect(() => PaymentIntent.new_(baseDraft())).not.toThrow();
  });

  test("rejects every field a decision would later depend on", () => {
    const cases: Array<Partial<ReturnType<typeof baseDraft>>> = [
      { x402Version: 1 },
      { scheme: "upto" },
      { networkCaip2: "solana:whatever" },
      { mint: "USDC" },
      { recipient: "not-base58!" },
      { atomicAmount: 0n },
      { authorizedResourceUrl: "/weather" },
      { authorizedResourceUrl: "file:///etc/passwd" },
      { httpMethod: "  " },
      { httpMethod: "GET /other HTTP/1.1" },
      { challengeHashHex: "abc" },
      { maxTimeoutSeconds: 0 },
      { maxTimeoutSeconds: 86401 },
      { identity: { ...identity(), tenantId: " " } },
      { identity: { ...identity(), requestId: "" } },
    ];
    for (const patch of cases) {
      expect(() => PaymentIntent.new_({ ...baseDraft(), ...patch })).toThrow();
    }
  });
});

describe("wire (de)serialization cannot bypass validation", () => {
  test("round-trips and rejects tampered payloads", () => {
    const original = intent("POST", enc.encode("payload"));
    const wire = original.toWire();
    expect(wire.atomic_amount).toBe(2500);
    const decoded = PaymentIntent.fromWire(JSON.parse(JSON.stringify(wire)));
    expect(decoded.equals(original)).toBe(true);
    expect(decoded.intentHashHex()).toBe(original.intentHashHex());

    const tamperedScheme = { ...wire, scheme: "upto" };
    expect(() => PaymentIntent.fromWire(tamperedScheme)).toThrow();
    const tamperedAmount = { ...wire, atomic_amount: 0 };
    expect(() => PaymentIntent.fromWire(tamperedAmount)).toThrow();
  });
});
